### ISSUE TITLE
Cacheable/RemoteCache does not work with new style classes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11038]
|<img alt="karlb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| karlb reported|
|-|-|
|Trac ID|trac#1324|
|Type|enhancement|
|Created|2005-11-11 03:14:49Z|

Attachments:

* [example.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f33/f3341dbcdda605b1601524b0d01655750cc60e0d/05e0372f8165776bed0a53b989eaffd9e3139848.py) (599 bytes) - added by karlb on 2005-11-11 03:14:49Z - 
* [cache-new-style-1324.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f33/f3341dbcdda605b1601524b0d01655750cc60e0d/645966e2f5daaaabcd16e1c1971778e3b1ce59dd.patch) (3418 bytes) - added by grudy on 2006-04-28 20:21:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1324 1324
type__enhancement enhancement
reporter__karlb karlb
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1131678889000000 1131678889000000
changetime__1308657168000000 1308657168000000
version__ 
owner__warner warner
cc__warner cc__karlb
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="karlb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|karlb commented|
|-|-|

```
#!html
<pre>
Passing new style classes which inherit from Cacheable throws an "TypeError: 
__class__ must be set to a class" exception.

A small example is attached. I'll try to write a real test over the weekend.

Expected output:
&lt;__main__.Data instance at 0x4a7418>

Real output:
Peer will receive following PB traceback:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/banana.py", line 114, in gotItem
    self.callExpressionReceived(item)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/banana.py", line 81, in callExpressionReceived
    self.expressionReceived(obj)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/pb.py", line 563, in expressionReceived
    method(*sexp[1:])
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/pb.py", line 874, in proto_message
    self._recvMessage(self.localObjectForID, requestID, objectID, message, 
answerRequired, netArgs, netKw)
--- &lt;exception caught here> ---
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/pb.py", line 888, in _recvMessage
    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/flavors.py", line 113, in remoteMessageReceived
    args = broker.unserialize(args)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/pb.py", line 824, in unserialize
    return unjelly(sexp, self.security, None, self)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/jelly.py", line 906, in unjelly
    return _Unjellier(taster, persistentLoad, invoker).unjellyFull(sexp)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/jelly.py", line 535, in unjellyFull
    o = self.unjelly(obj)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/jelly.py", line 571, in unjelly
    ret = thunk(obj[1:])
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/jelly.py", line 654, in _unjelly_tuple
    if isinstance(self.unjellyInto(l, elem, lst[elem]), NotKnown):
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/jelly.py", line 621, in unjellyInto
    o = self.unjelly(jel)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/jelly.py", line 558, in unjelly
    val = method(self, obj)
  File "/Library/Frameworks/Python.framework/Versions/2.4/lib/python2.4/site-
packages/twisted/spread/flavors.py", line 450, in unjellyFor
    cProxy.__class__ = self.__class__
exceptions.TypeError: __class__ must be set to a class
Unhandled error in Deferred:
Traceback from remote host -- Traceback unavailable
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="grudy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|grudy commented|
|-|-|

Patch attached

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `closed`|
|-|-|

(In [17167]) Support new style classes for cacheables

New-style copyables have long been supported.  Here, we rectify the oversight
that cacheables were not updated to make new-style classes jellyable over PB as
well.

Author: karlb

Reviewer: glyph

Fixes [#11038](https://github.com/twisted/twisted/issues/11038)
